src/System.Private.Reflection.Core/src/System/Reflection/Runtime/General/EcmaFormat/ReflectionTypeProvider.cs

```diff
@@ -0,0 +1,288 @@
+                    return CommonRuntimeTypes.Double;
+                case PrimitiveTypeCode.Int16:
+                    return CommonRuntimeTypes.Int16;
+                case PrimitiveTypeCode.Int32:
+                    return CommonRuntimeTypes.Int32;
+                case PrimitiveTypeCode.Int64:
+                    return CommonRuntimeTypes.Int64;
+                case PrimitiveTypeCode.IntPtr:
+                    return CommonRuntimeTypes.IntPtr;
+                case PrimitiveTypeCode.Object:
+                    return CommonRuntimeTypes.Object;
+                case PrimitiveTypeCode.SByte:
+                    return CommonRuntimeTypes.SByte;
+                case PrimitiveTypeCode.Single:
+                    return CommonRuntimeTypes.Single;
+                case PrimitiveTypeCode.String:
+                    return CommonRuntimeTypes.String;
+                case PrimitiveTypeCode.TypedReference:
+                    throw new PlatformNotSupportedException();
+                case PrimitiveTypeCode.UInt16:
+                    return CommonRuntimeTypes.UInt16;
+                case PrimitiveTypeCode.UInt32:
+                    return CommonRuntimeTypes.UInt32;
+                case PrimitiveTypeCode.UInt64:
+                    return CommonRuntimeTypes.UInt64;
+                case PrimitiveTypeCode.UIntPtr:
+                    return CommonRuntimeTypes.UIntPtr;
+                case PrimitiveTypeCode.Void:
+                    return CommonRuntimeTypes.Void;
+            }
+
+            throw new BadImageFormatException();
+        }
+    }
+    class ReflectionTypeProvider : ICustomAttributeTypeProvider<RuntimeTypeInfo>, ISZArrayTypeProvider<RuntimeTypeInfo>, ISignatureTypeProvider<RuntimeTypeInfo, TypeContext>
+    {
+        private static readonly RuntimeTypeInfo s_systemType = (RuntimeTypeInfo)CommonRuntimeTypes.Type.GetTypeInfo();
+        private static readonly RuntimeTypeInfo s_intPtrType = (RuntimeTypeInfo)CommonRuntimeTypes.IntPtr.GetTypeInfo();
+
+        private bool _throwOnError;
+        private Exception _exceptionResult;
+        public bool ExceptionOccurred { get {return _exceptionResult != null; } }
+        public Exception ExceptionResult
+        {
+            get
+            {
+                return _exceptionResult;
+            }
+            set
+            {
+                if (_exceptionResult == null)
+                {
+                    _exceptionResult = value;
+
+                    if (_exceptionResult != null)
+                        throw _exceptionResult;
+                }
+            }
+        }
+
+        public ReflectionTypeProvider(bool throwOnError)
+        {
+            _throwOnError = throwOnError;
+        }
+
+        RuntimeTypeInfo ICustomAttributeTypeProvider<RuntimeTypeInfo>.GetSystemType()
+        {
+            return s_systemType;
+        }
+
+        bool ICustomAttributeTypeProvider<RuntimeTypeInfo>.IsSystemType(RuntimeTypeInfo type)
+        {
```

Comment:
BTW, you don't need to cast System.Type to a TypeInfo (or RuntimeTypeInfo) just to compare it to another type. That'll probably eliminate the need to cache s_systemType altogether.